multi-era/rust/src/byron/block/serialization.rs

```diff
@@ -0,0 +1,1019 @@
+                })
+                .map_err(|e: DeserializeError| e.annotate("extra_proof"))?;
+            match len {
+                cbor_event::Len::Len(_) => (),
+                cbor_event::Len::Indefinite => match raw.special()? {
+                    cbor_event::Special::Break => (),
+                    _ => return Err(DeserializeFailure::EndingBreakMissing.into()),
+                },
+            }
+            Ok(BlockHeaderExtraData {
+                block_version,
+                software_version,
+                byron_attributes,
+                extra_proof,
+            })
+        })()
+        .map_err(|e| e.annotate("BlockHeaderExtraData"))
+    }
+}
+
+impl cbor_event::se::Serialize for ByronBlock {
+    fn serialize<'se, W: Write>(
+        &self,
+        serializer: &'se mut Serializer<W>,
+    ) -> cbor_event::Result<&'se mut Serializer<W>> {
+        match self {
+            ByronBlock::EpochBoundary(epoch_boundary) => epoch_boundary.serialize(serializer),
+            ByronBlock::Main(main) => main.serialize(serializer),
+        }
+    }
+}
+
+impl Deserialize for ByronBlock {
+    fn deserialize<R: BufRead + Seek>(raw: &mut Deserializer<R>) -> Result<Self, DeserializeError> {
+        (|| -> Result<_, DeserializeError> {
+            let initial_position = raw.as_mut_ref().seek(SeekFrom::Current(0)).unwrap();
+            let mut errs = Vec::new();
+            let deser_variant: Result<_, DeserializeError> = ByronEbBlock::deserialize(raw);
+            match deser_variant {
+                Ok(epoch_boundary) => return Ok(Self::EpochBoundary(epoch_boundary)),
+                Err(e) => {
+                    errs.push(e.annotate("EpochBoundary"));
+                    raw.as_mut_ref()
+                        .seek(SeekFrom::Start(initial_position))
+                        .unwrap();
+                }
+            };
+            let deser_variant: Result<_, DeserializeError> = ByronMainBlock::deserialize(raw);
+            match deser_variant {
+                Ok(main) => return Ok(Self::Main(main)),
+                Err(e) => {
+                    errs.push(e.annotate("Main"));
+                    raw.as_mut_ref()
+                        .seek(SeekFrom::Start(initial_position))
+                        .unwrap();
+                }
+            };
+            Err(DeserializeError::new(
+                "ByronBlock",
+                DeserializeFailure::NoVariantMatchedWithCauses(errs),
+            ))
+        })()
+        .map_err(|e| e.annotate("ByronBlock"))
+    }
+}
+
+impl cbor_event::se::Serialize for ByronBlockBody {
+    fn serialize<'se, W: Write>(
+        &self,
+        serializer: &'se mut Serializer<W>,
+    ) -> cbor_event::Result<&'se mut Serializer<W>> {
+        serializer.write_array(cbor_event::Len::Len(4))?;
+        serializer.write_array(cbor_event::Len::Indefinite)?;
+        //serializer.write_array(cbor_event::Len::Len(self.tx_payload.len() as u64))?;
```

Comment:
why we needed to change this?

Reply:
See the discussion in the tooling channel. Byron encodes a few areas as always indefinite but the rest is canonical CBOR. Before we thought it was 100% canonical